### ISSUE TITLE
Enhance conflict error messages with the dependency chains that lead to the conflict 

### ISSUE THREAD

**Issue by syslaila** [mamba-org/mamba#1111]
When a conflict happens/or a package cannot be installed because one of the dependencies doesn't have any provider, mamba/libsolv would only show the specification of the package that cannot be installed.
It would be useful, however, to show what introduces the conflict and what is the chain of dependencies that lead to a conflict.

For example:
A user requests packages A (depends on package C 1.0.0) and B (depends on package C 2.0.0). This leads to a conflict which shows:
"Problem: package A requires C 1.0.0, but none of the providers can be installed"

It would be useful to also show that we requested 2 different versions of C and who required them (especially useful if the chain of dependencies up to the C library is very complicated A -> -> C 1.0.0, same with B).

Is it possible to somehow use the "whoneeds" functionality to retrieve this information ? As far as I see this currently works for an installed environment.

opened an equivalent issue on libsolv: https://github.com/openSUSE/libsolv/issues/467.

Happy to contribute for this functionality :)

**Comment by adriendelsalle** [comment]
Thanks a lot @syslaila for opening this issue!

We would be super happy to have contributions on that. And you're right `libsolv` is probably the right place to work on this.
Let's us know if we can provide some guidance, we plan to work on this topic in a mid-term time frame but we could accelerate a bit the agenda if you have bandwidth :).

**Comment by wolfv** [comment]
Hi @syslaila "whoneeds" works with channels as well. E.g. you can do `mamba repoquery whoneeds xtensor -c conda-forge`.

however, I don't think that's sufficient to get to the information you want. I think we need to inspect the intermediate solver state, and I also don't know exactly how we should do this.

There is another solver called "pubgrub" that claims to produce nice error messages.

An implementation in Rust can be found here: https://github.com/pubgrub-rs/pubgrub
There is additional docs here https://github.com/dart-lang/pub/blob/master/doc/solver.md and https://nex3.medium.com/pubgrub-2fb6470504f

It could be interesting to try to feed pubgrub-rs with conda metadata and check what kind of error messages we can get from it. For libsolv, you can inspect the output from the backtracking process when you pass `-vvv` to mamba. We need to figure out how to condense this output to meaningful information.

**Comment by syslaila** [comment]
hey @wolfv, 
I saw that improving the error messaging is tracked based on: https://wolfv.medium.com/the-mamba-project-and-the-czi-grant-ec88fb27c25, this is very exciting!
Unfortunately, I didn't have bandwidth to look into this the past months but I would have some bandwidth to work on this topic in ~ 3 weeks. I was wondering if this work has started already and what would be next steps ? intergrating the conda metadata with pubgrub or checking with libsolv how we can track the chain of dependencies for a solve conflict ?   
I checked a little bit the pubgrub solver, they seems to solve the environment using version ranges and I don't see anything about supporting union of ranges eg. 1.0|2.0 (https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#package-match-specifications)
